tephra-core/src/main/java/co/cask/tephra/persist/HDFSTransactionLog.java

```diff
@@ -107,45 +119,100 @@ public TransactionLogReader getReader() throws IOException {
     return reader;
   }
 
-  private static final class LogWriter implements TransactionLogWriter {
-    private final SequenceFile.Writer internalWriter;
+  @VisibleForTesting
+  static byte[] toBytes(int val) {
+    byte [] b = new byte[4];
+    for (int i = 3; i > 0; i--) {
+      b[i] = (byte) val;
+      val >>>= 8;
+    }
+    b[0] = (byte) val;
+    return b;
+  }
 
+  @VisibleForTesting
+  static final class LogWriter implements TransactionLogWriter {
+    private final SequenceFile.Writer internalWriter;
+    private int numEntries;
+    private List<Entry> transactionEntries;
     public LogWriter(FileSystem fs, Configuration hConf, Path logPath) throws IOException {
       // TODO: retry a few times to ride over transient failures?
+      SequenceFile.Metadata metadata = new SequenceFile.Metadata();
+      metadata.set(new Text("version"), new Text("v2"));
+
       this.internalWriter =
-        SequenceFile.createWriter(fs, hConf, logPath, LongWritable.class, TransactionEdit.class);
+        SequenceFile.createWriter(fs, hConf, logPath, LongWritable.class,
+                                  TransactionEdit.class, SequenceFile.CompressionType.NONE, null, null, metadata);
+      numEntries = 0;
+      transactionEntries = new ArrayList<>();
+
       LOG.debug("Created a new TransactionLog writer for " + logPath);
     }
 
     @Override
     public void append(Entry entry) throws IOException {
-      internalWriter.append(entry.getKey(), entry.getEdit());
+      transactionEntries.add(entry);
```

Comment:
Instead of collecting transaction edit logs here, we can add the count header in `AbstractTransactionLog.sync()` method before calling append, right? Since the edit logs are already being collected over there, we will then avoid code duplication.